lib/ansible/module_utils/csharp/Ansible.Basic.cs

```diff
@@ -242,10 +243,18 @@ public void Debug(string message)
                 LogEvent(String.Format("[DEBUG] {0}", message));
         }
 
-        public void Deprecate(string message, string version)
+        public void Deprecate(string message, string version = null, string date = null)
```

Comment:
Unfortunately PowerShell does not handle kwargs in C# methods very well, we cannot do `$module.Deprecate('msg', date='date')` we would have to do `$module.Deprecate('msg', $null, 'date')` which isn't ideal. An alternative is to create an overloaded method

```
public void Deprecate(string message, string version)
{
    Existing logic
}

public void Deprecate(string message, DateTime date)
{
    New logic
}
```

That way a module can differentiate between the 2 with `$module.Deprecate('msg', [DateTime]'2020-03-27')  or use whatever other date format it wants and the C# method can format that date into a common output format, preferably ISO 8601 `yyyy-mm-dd`.

Reply:
How do I convert a `DateTime` object to ISO 8601? (Sorry for asking, I had to write that code essentially blind since I have neither C# nor PowerShell available for testing...)

Reply:
No need to apologise for asking :) A DateTime object's [ToString](https://docs.microsoft.com/en-us/dotnet/api/system.datetime.tostring?view=netframework-4.8#System_DateTime_ToString_System_String_) method can take in a string that provides how to format the date as a string. In this case you want to do `date.ToString("yyyy-MM-dd")`.

The format can be either a known [standard format](https://docs.microsoft.com/en-us/dotnet/standard/base-types/standard-date-and-time-format-strings?view=netframework-4.8) or a [custom format](https://docs.microsoft.com/en-us/dotnet/standard/base-types/custom-date-and-time-format-strings?view=netframework-4.8).